### PR TITLE
BUG: read_csv not recognizing numbers appropriately when decimal is set

### LINKED ISSUE #31920: pd.read_csv does not recognize scientific notation if 'decimal' attribute is set with engine=python
Contents of 'abc.tsv':
```
a	b
1,2	3,5E-1
1,2e2	3,5
```
Code:
```python
>>> import pandas as pd
>>> pd.read_csv('abc.tsv', '\t', decimal=',')
       a     b
0    1.2  0.35
1  120.0  3.50
>>> pd.read_csv('abc.tsv', '\t', decimal=',', engine='python')
       a       b
0    1.2  3,5E-1
1  1,2e2     3.5
```
#### Problem description
The outputs of the two read_csv commands above should be the same. In this case an obvious "solution" is to just use the C-engine, however I have to process a file where I need to use a regular expression as the delimiter.

#### Expected Output
```python
>>> pd.read_csv('abc.tsv', '\t', decimal=',', engine='python')
       a     b
0    1.2  0.35
1  120.0  3.50
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.2.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : None
LOCALE           : en_US.UTF-8

pandas           : 1.0.1
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.2.0.post20200210
Cython           : 0.29.15
pytest           : 5.3.5
hypothesis       : 5.4.1
sphinx           : 2.4.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.12.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.1.3
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.13
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.7
numba            : 0.48.0

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #31920
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The old regex was shorter but pretty buggy.

Let's assume ``decimal=","`` and ``thousands="."``

For example something like ``1a.2,3`` was interpreted as numeric by the regex and converted to ``1a2.3``, because the ``.`` was not escaped in the regex.
Also something like ``1,2,3`` was interpreted as numeric by the regex and converted to ``1.2.3``. 

This one is not quite finished. We have to define a few thousand separator related things:

How strict would we want to be? Should ``1.2,3`` be interpreted as numeric and be converted to ``12.3`` or is only something like ``1.234,5`` relevant as thousands separator? C Engine validation is not strict
Additionally should ``,2`` be the number ``0.2``? -> Currently it is, because the C engine behaves the same

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -2346,9 +2346,17 @@ def __init__(self, f: Union[FilePathOrBuffer, List], **kwds):
             raise ValueError("Only length-1 decimal markers supported")
 
         if self.thousands is None:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.decimal}]+")
+            regex = fr"^\-?[0-9]*({self.decimal}[0-9]*)?([0-9](E|e)\-?[0-9]*)?$"
```
Comment (jbrockmendel) [inline pandas/io/parsers.py:2349]: is self.decimal already escaped by the time we get here?

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -3036,7 +3044,7 @@ def _search_replace_num_columns(self, lines, search, replace):
                     not isinstance(x, str)
                     or search not in x
                     or (self._no_thousands_columns and i in self._no_thousands_columns)
-                    or self.nonnum.search(x.strip())
+                    or not self.nonnum.search(x.strip())
```
Comment (jbrockmendel) [inline pandas/io/parsers.py:3047]: this flips the meaning of nonnum right?  should rename?

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -2346,9 +2346,17 @@ def __init__(self, f: Union[FilePathOrBuffer, List], **kwds):
...
-            self.nonnum = re.compile(fr"[^-^0-9^{self.decimal}]+")
+            regex = fr"^\-?[0-9]*({self.decimal}[0-9]*)?([0-9](E|e)\-?[0-9]*)?$"
+            self.nonnum = re.compile(regex)
         else:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.thousands}^{self.decimal}]+")
+            thousands = self.thousands
+            if "." == thousands:
+                thousands = fr"\{thousands}"
+            regex = (
+                fr"^\-?([0-9]+{thousands}|[0-9])*({self.decimal}[0-9]*)?"
```
Comment (jbrockmendel) [inline pandas/io/parsers.py:2356]: might be out of scope, but do we want to restrict to 3 digits between thousands characters (IIRC in some locales they separate on ten-thousands, not sure if we support that)

Comment (phofl) [reply pandas/io/parsers.py:2356]: That was a thing I was wondering too when opening this PR. Not quite sure how strict we would want to be here. We could maybe check for at least three digits between? The old version did not check anything related to this

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -2346,9 +2346,17 @@ def __init__(self, f: Union[FilePathOrBuffer, List], **kwds):
             raise ValueError("Only length-1 decimal markers supported")
 
         if self.thousands is None:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.decimal}]+")
+            regex = fr"^\-?[0-9]*({self.decimal}[0-9]*)?([0-9](E|e)\-?[0-9]*)?$"
+            self.nonnum = re.compile(regex)
         else:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.thousands}^{self.decimal}]+")
+            thousands = self.thousands
+            if "." == thousands:
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/io/parsers.py:2353]: Actually this raises a good point. Independently of the expected input (we do not say anything about this in the docs), maybe we should escape decimals too?

File: pandas/io/parsers.py (hunk 2 above)
Comment (phofl) [reply pandas/io/parsers.py:3047]: Yep is better, done

File: pandas/io/parsers.py (hunk 1 above)
Comment (phofl) [reply pandas/io/parsers.py]: No, seems not to be the case. Escaped it.

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 5]
```diff
@@ -314,3 +314,51 @@ def test_malformed_skipfooter(python_parser_only):
...
+)
+def test_decimal_and_exponential_erroneous(python_parser_only, thousands, value):
+    # GH#31920
+    data = StringIO(
+        f"""a	b
+    1,1	{value}
+    """
+    )
+    result = python_parser_only.read_csv(
+        data, "\t", decimal=",", engine="python", thousands=thousands
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_python_parser_only.py:361]: * `engine="python"` is extraneous

* How well supported is this in the `c` engine?

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 6]
```diff
@@ -314,3 +314,51 @@ def test_malformed_skipfooter(python_parser_only):
...
+)
+def test_decimal_and_exponential(python_parser_only, thousands, value, result_value):
+    # GH#31920
+    data = StringIO(
+        f"""a	b
+    1,1	{value}
+    """
+    )
+    result = python_parser_only.read_csv(
+        data, "\t", decimal=",", engine="python", thousands=thousands
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_python_parser_only.py:342]:
```suggestion
        data, "\t", decimal=",", thousands=thousands
```

Comment (phofl) [reply pandas/tests/io/parser/test_python_parser_only.py:333]: Sorry wrong commit button above. C works perfectly here. Already have tests therefore. Would probably makes sense unifying them as a follow up

Comment (gfyoung) [reply pandas/tests/io/parser/test_python_parser_only.py:333]: IMO would do it in this PR, but follow-up also works.

File: doc/source/whatsnew/v1.3.0.rst [hunk 7]
```diff
@@ -255,6 +255,7 @@ I/O
 ^^^
 
 - Bug in :meth:`Index.__repr__` when ``display.max_seq_items=1`` (:issue:`38415`)
+- Bug in :func:`read_csv` not recognizing scientific notation if decimal is set (:issue:`31920`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:258]: this should say engine='python' right?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 7 above)
Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst:258]: Yep

Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst]: Changed

File: pandas/tests/io/parser/test_python_parser_only.py (hunk 6 above)
Comment (jreback) [reply pandas/tests/io/parser/test_python_parser_only.py:333]: this is fine as a followup

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
great ping on green (suggestion to combine the python/c tests as a followup)

**Comment by phofl** [comment]
@jreback green

**Comment by jreback** [comment]
thanks @phofl if you can do a PR or issue for the followup would be great

**Comment by phofl** [comment]
Will open an issue. Hopefully I will be able to get back to this in the coming days

**Comment by phofl** [comment]
Opened #38926

**Pull request merged**